File: impls/rust/src/step0_repl.rs
use rustyline::Editor;

fn main() {
    let mut rl = Editor::<()>::new();
    rl.load_history("history.txt").ok();
    while let Ok(line) = rl.readline("user> ") {
        rl.add_history_entry(&line);

        let result = print(eval(read(line)));
        print!("{}", result);
    }
    rl.save_history("history.txt").unwrap();
}

fn read(input: String) -> String {
    input
}

fn eval(input: String) -> String {
    input
}

fn print(input: String) -> String {
    input
}

File: impls/js2/step0_repl.mjs
import { createInterface } from "readline";

const rl = createInterface({
  input: process.stdin,
  output: process.stdout,
  terminal: true,
});

function js_eval(fn_body) {
  return Function(`"use strict";${fn_body}`)();
}

function log(value) {
  console.log(value);
  return value;
}

rl.setPrompt("user> ");
rl.prompt();
rl.on("line", (line) => {
  console.log(compiled_rep(line));
  rl.prompt();
});
rl.on("close", () => {
  console.log();
});

function READ(input) {
  return input;
}
function EVAL(input) {
  return input;
}
function PRINT(input) {
  return input;
}
function rep(input) {
  return PRINT(EVAL(READ(input)));
}

function compile(input) {
  return (
    'return "' + input.replaceAll("\\", "\\\\").replaceAll('"', '\\"') + '";'
  );
}
function compiled_rep(input) {
  return PRINT(js_eval(compile(READ(input))));
}
